pageserver/src/tenant/remote_timeline_client.rs

```diff
@@ -1713,17 +1728,44 @@ impl RemoteTimelineClient {
                     upload_queue.num_inprogress_layer_uploads -= 1;
                     None
                 }
-                UploadOp::UploadMetadata(_, lsn) => {
+                UploadOp::UploadMetadata { ref uploaded } => {
                     upload_queue.num_inprogress_metadata_uploads -= 1;
-                    // XXX monotonicity check?
 
-                    upload_queue.projected_remote_consistent_lsn = Some(lsn);
-                    if self.generation.is_none() {
-                        // Legacy mode: skip validating generation
-                        upload_queue.visible_remote_consistent_lsn.store(lsn);
-                        None
+                    // the task id is reused as a monotonicity check for storing the "clean"
+                    // IndexPart.
+                    let last_updater = upload_queue.clean.1;
+                    let is_later = last_updater.is_some_and(|task_id| task_id < task.task_id);
+                    let monotone = is_later || last_updater.is_none();
+
+                    if monotone {
+                        // not taking ownership is wasteful
+                        upload_queue.clean.0.clone_from(uploaded);
+                        upload_queue.clean.1 = Some(task.task_id);
+
+                        let lsn = upload_queue.clean.0.metadata.disk_consistent_lsn();
+
+                        if self.generation.is_none() {
+                            // Legacy mode: skip validating generation
+                            upload_queue.visible_remote_consistent_lsn.store(lsn);
+                            None
+                        } else {
+                            Some((lsn, upload_queue.visible_remote_consistent_lsn.clone()))
+                        }
                     } else {
-                        Some((lsn, upload_queue.visible_remote_consistent_lsn.clone()))
+                        // do we want to log? this is perfectly valid, since the ordering in
```

Comment:
Since we wait for previous index uploads before doing an index upload, does that mean this branch can never be taken?  (i.e. we would never have racing completion of index uploads).  If we can figure that out, then this can be an assertion rather than a rarely taken brancch.

Reply:
[a2463ca](https://github.com/neondatabase/neon/pull/7833/commits/a2463ca0753df56a4f238a5bee61635d47543605)